src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LogValuesFormatter.cs

```diff
@@ -29,35 +33,42 @@ public LogValuesFormatter(string format)
 
             OriginalFormat = format;
 
-            var sb = new StringBuilder();
+            var vsb = new ValueStringBuilder(stackalloc char[256]);
             int scanIndex = 0;
             int endIndex = format.Length;
 
             while (scanIndex < endIndex)
             {
                 int openBraceIndex = FindBraceIndex(format, '{', scanIndex, endIndex);
+                if (scanIndex == 0 && openBraceIndex == endIndex)
+                {
+                    // No holes found.
+                    _format = format;
+                    return;
+                }
+
                 int closeBraceIndex = FindBraceIndex(format, '}', openBraceIndex, endIndex);
 
                 if (closeBraceIndex == endIndex)
                 {
-                    sb.Append(format, scanIndex, endIndex - scanIndex);
+                    vsb.Append(format.AsSpan(scanIndex, endIndex - scanIndex));
                     scanIndex = endIndex;
                 }
                 else
                 {
                     // Format item syntax : { index[,alignment][ :formatString] }.
                     int formatDelimiterIndex = FindIndexOfAny(format, FormatDelimiters, openBraceIndex, closeBraceIndex);
 
-                    sb.Append(format, scanIndex, openBraceIndex - scanIndex + 1);
-                    sb.Append(_valueNames.Count.ToString(CultureInfo.InvariantCulture));
+                    vsb.Append(format.AsSpan(scanIndex, openBraceIndex - scanIndex + 1));
+                    vsb.Append(_valueNames.Count.ToString());
                     _valueNames.Add(format.Substring(openBraceIndex + 1, formatDelimiterIndex - openBraceIndex - 1));
```

Comment:
Would it be worth delaying this Substring alloc?

Reply:
Delaying it to when?

Reply:
Delaying it until the string is actually used.

https://source.dot.net/#Microsoft.Extensions.Logging.Abstractions/LogValuesFormatter.cs,a31f3c3b2cea5615,references

I am not super familiar with this area, but if we only captured the Ranges here, and returned the `_valueNames` as `ReadOnlyMemory<char>`, maybe we could save alloc'ing these as `string`s all together?

Reply:
> maybe we could save alloc'ing these as strings all together?

If this is likely, then yeah, we should consider that.  But I have no idea how common that is (and if it's at all common, it'll end up being worse as we'll end up having stored the ranges and then needing to also store / replace that with the constructed strings).

@davidfowl?

Let's address it separately.

Reply:
I would stringy those up front.

Reply:
Meaning as-is, right? Does that mean these names are generally used?

Reply:
Yep. We have another source generator thing that will replace this.

Reply:
This was roughly what I was thinking: https://github.com/eerhardt/runtime/commit/aa3345733a9fad585c70c05fad93f7b3d796528b. At first I thought I could eliminate the substring allocs all together.

However, it appears that there is a hidden expectation that the scope objects implement `IReadOnlyList<KeyValuePair<string, object>>`, which means this would be a breaking change. Or we could make the `LogValues` structs implement both `IReadOnlyList<KeyValuePair<string, object>>` and ``IReadOnlyList<KeyValuePair<ReadOnlyMemory<char>, object>>`, and have the Console logger (and any other loggers) check for both.